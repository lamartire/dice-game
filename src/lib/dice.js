import {
  isEqual,
  random,
  times,
  find,
  pick,
  sum,
  map,
  flatMap,
} from 'lodash/fp'

/**
 *
 */
export class Dice {
  constructor(rounds = 5) {
    this.rounds = rounds
    this.round = 1
    this.players = []
    this.currentPlayer = null
  }

  get state() {
    return pick(['players', 'round', 'currentPlayer', 'winner'], this)
  }

  get winner() {
    const [playerA, playerB] = this.players
    const isGameFinished = this.round > this.rounds

    if (!isGameFinished) {
      return null
    }

    const playerAScore = this.getThrowsSum(playerA.throws)
    const playerBScore = this.getThrowsSum(playerB.throws)

    if (playerAScore > playerBScore) {
      return {
        ...playerA,
        score: playerAScore,
      }
    }

    return {
      ...playerB,
      score: playerBScore,
    }
  }

  getThrowsSum(throws) {
    return sum(flatMap(values => sum(values), throws))
  }

  createPlayer(name) {
    return {
      name: name,
      throws: [],
    }
  }

  addPlayer(name) {
    if (this.players.length === 2) {
      throw new Error('Only two players can playing at one time')
    } else if (find(player => isEqual(player.name, name), this.players)) {
      throw new Error(`Player with ${name} already added!`)
    }

    const newPlayer = this.createPlayer(name)

    if (!this.currentPlayer) {
      this.currentPlayer = newPlayer
    }

    this.players.push(newPlayer)
  }

  throwDices() {
    return times(() => random(1, 6), 2)
  }

  moveQueue() {
    const [playerA, playerB] = this.players

    if (isEqual(this.currentPlayer, playerA)) {
      this.currentPlayer = playerB
    } else {
      this.currentPlayer = playerA
      this.round += 1
    }
  }

  turn() {
    const result = this.throwDices()

    this.currentPlayer.throws.push(result)
    this.moveQueue()

    return this.state
  }
}
