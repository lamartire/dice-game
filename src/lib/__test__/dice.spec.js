import { times } from 'lodash/fp'
import { Dice } from '../dice'

describe('Dice game', () => {
  describe('initialization and instance specification', () => {
    it('creates instance of Dice', () => {
      expect(new Dice()).toBeInstanceOf(Dice)
    })

    it('by default creates with 5 rounds', () => {
      expect(new Dice().rounds).toMatchSnapshot()
    })
  })

  describe('common methods', () => {
    const dice = new Dice()

    it('creates new player with given name', () => {
      expect(dice.createPlayer('Foo')).toMatchSnapshot()
    })

    it('throws dices and returns result', () => {
      expect(dice.throwDices()).toEqual([
        expect.any(Number),
        expect.any(Number),
      ])
    })
  })

  describe('game process', () => {
    it('creates new player with given name and stores it', () => {
      const dice = new Dice()

      dice.addPlayer('Foo')

      expect(dice.players).toMatchSnapshot()
    })

    it('throws if already added two players', done => {
      try {
        const dice = new Dice()

        dice.addPlayer('Foo')
        dice.addPlayer('Bar')
        dice.addPlayer('Baz')
      } catch (err) {
        done()
      }
    })

    it('throws error if player with passed name already added', done => {
      try {
        const dice = new Dice()

        dice.addPlayer('Foo')
        dice.addPlayer('Foo')
      } catch (err) {
        done()
      }
    })

    it('after adding first player makes him as current', () => {
      const dice = new Dice()

      expect(dice.currentPlayer).toBeNull()

      dice.addPlayer('Foo')

      expect(dice.currentPlayer).not.toBeNull()
      expect(dice.currentPlayer).toMatchSnapshot()
    })

    it('move queue to next player, after last player increments round', () => {
      const dice = new Dice()

      dice.addPlayer('Foo')

      const lastPlayer = { ...dice.currentPlayer }

      dice.addPlayer('Bar')
      dice.moveQueue()

      expect(dice.currentPlayer).not.toMatchObject(lastPlayer)

      dice.moveQueue()

      expect(dice.currentPlayer).toMatchObject(lastPlayer)
      expect(dice.round).toEqual(2)
    })

    it('adds result of throwing to currentPlayer and makes as current next player', () => {
      const dice = new Dice()

      dice.addPlayer('Foo')
      dice.addPlayer('Bar')
      dice.turn()

      const [player1, player2] = dice.players

      expect(dice.currentPlayer).toMatchObject(player2)
      expect(player1.throws).toHaveLength(1)
    })

    it('launches next round after throw of last player', () => {
      const dice = new Dice()

      dice.addPlayer('Foo')
      dice.addPlayer('Bar')
      dice.turn()
      dice.turn()

      expect(dice.round).toEqual(2)
    })

    it('stops game after throw of last player in final round and marks someone as winner', () => {
      const dice = new Dice()

      dice.addPlayer('Foo')
      dice.addPlayer('Bar')
      times(() => dice.turn(), 5 * 2)

      expect(dice.winner).not.toBeNull()
    })
  })
})
