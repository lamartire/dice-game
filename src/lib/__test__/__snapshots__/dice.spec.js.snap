// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dice game common methods creates new player with given name 1`] = `
Object {
  "name": "Foo",
  "throws": Array [],
}
`;

exports[`Dice game game process after adding first player makes him as current 1`] = `
Object {
  "name": "Foo",
  "throws": Array [],
}
`;

exports[`Dice game game process creates new player with given name and stores it 1`] = `
Array [
  Object {
    "name": "Foo",
    "throws": Array [],
  },
]
`;

exports[`Dice game game process stops game after throw of last player in final round and marks someone as winner 1`] = `
Object {
  "currentPlayer": Object {
    "name": "Foo",
    "throws": Array [
      Array [
        1,
        4,
      ],
      Array [
        4,
        4,
      ],
      Array [
        4,
        2,
      ],
      Array [
        3,
        5,
      ],
      Array [
        1,
        5,
      ],
    ],
  },
  "players": Array [
    Object {
      "name": "Foo",
      "throws": Array [
        Array [
          1,
          4,
        ],
        Array [
          4,
          4,
        ],
        Array [
          4,
          2,
        ],
        Array [
          3,
          5,
        ],
        Array [
          1,
          5,
        ],
      ],
    },
    Object {
      "name": "Bar",
      "throws": Array [
        Array [
          1,
          4,
        ],
        Array [
          4,
          2,
        ],
        Array [
          6,
          1,
        ],
        Array [
          5,
          5,
        ],
        Array [
          6,
          1,
        ],
      ],
    },
  ],
  "round": 6,
  "winner": Object {
    "name": "Bar",
    "throws": Array [
      Array [
        1,
        4,
      ],
      Array [
        4,
        2,
      ],
      Array [
        6,
        1,
      ],
      Array [
        5,
        5,
      ],
      Array [
        6,
        1,
      ],
    ],
  },
}
`;

exports[`Dice game initialization and instance specification by default creates with 5 rounds 1`] = `5`;
